go-controller/pkg/observability/observability.go

```diff
@@ -0,0 +1,210 @@
+	return nil
+}
+
+// Cleanup must be called when observability is no longer needed.
+// It will return an error if some samples still exist in the db.
+// This is expected, and Cleanup may be retried on the next restart.
+func Cleanup(nbClient libovsdbclient.Client) error {
+	// Do the opposite of init
+	err := libovsdbops.DeleteSamplingAppsWithPredicate(nbClient, func(app *nbdb.SamplingApp) bool {
+		return true
+	})
+	if err != nil {
+		return fmt.Errorf("error deleting sampling apps: %w", err)
+	}
+
+	err = libovsdbops.DeleteSampleCollectorWithPredicate(nbClient, func(collector *nbdb.SampleCollector) bool {
+		return true
+	})
+	if err != nil {
+		return fmt.Errorf("error deleting sample collectors: %w", err)
+	}
+	return nil
+}
+
+func (m *Manager) setSamplingAppIDs() error {
+	var ops []libovsdb.Operation
+	var err error
+	for _, appConfig := range []struct {
+		id      int
+		appType nbdb.SamplingAppType
+	}{
+		{
+			id:      DropSamplingID,
+			appType: nbdb.SamplingAppTypeDrop,
+		},
+		{
+			id:      ACLNewTrafficSamplingID,
+			appType: nbdb.SamplingAppTypeACLNew,
+		},
+		{
+			id:      ACLEstTrafficSamplingID,
+			appType: nbdb.SamplingAppTypeACLEst,
+		},
+	} {
+		samplingApp := &nbdb.SamplingApp{
+			ID:   appConfig.id,
+			Type: appConfig.appType,
+		}
+		ops, err = libovsdbops.CreateOrUpdateSamplingAppsOps(m.nbClient, ops, samplingApp)
+		if err != nil {
+			return fmt.Errorf("error creating or updating sampling app %s: %w", appConfig.appType, err)
+		}
+	}
+	_, err = libovsdbops.TransactAndCheck(m.nbClient, ops)
+	return err
+}
+
+func groupByProbability(c *collectorConfig) map[int][]libovsdbops.SampleFeature {
+	probabilities := make(map[int][]libovsdbops.SampleFeature)
+	for feature, percentProbability := range c.featuresProbability {
+		probability := percentToProbability(percentProbability)
+		probabilities[probability] = append(probabilities[probability], feature)
+	}
+	return probabilities
+}
+
+func getCollectorKey(collectorID int, probability int) string {
+	return fmt.Sprintf("%d-%d", collectorID, probability)
+}
+
+func (m *Manager) getFreeCollectorID() (int, error) {
+	for i := 1; i <= maxCollectorID; i++ {
+		if !m.takenCollectorIDs.Has(i) {
+			return i, nil
+		}
+	}
+	return 0, fmt.Errorf("no free collector IDs")
+}
+
+func (m *Manager) addCollector(conf *collectorConfig) (map[libovsdbops.SampleFeature][]string, error) {
+	sampleFeaturesConfig := make(map[libovsdbops.SampleFeature][]string)
+	probabilityConfig := groupByProbability(conf)
+
+	for probability, features := range probabilityConfig {
+		collectorKey := getCollectorKey(conf.collectorSetID, probability)
+		var collectorUUID string
+		var ok bool
+		if collectorUUID, ok = m.dbCollectors[collectorKey]; !ok {
+			collectorID, err := m.getFreeCollectorID()
+			if err != nil {
+				return sampleFeaturesConfig, err
+			}
+			collector := &nbdb.SampleCollector{
+				ID:          collectorID,
+				SetID:       conf.collectorSetID,
+				Probability: probability,
```

Comment:
`NB.Sample_Collector` has an `external_ids` column.  Does it make sense to store something there?  Maybe something to identify the features that require that probability

Reply:
To clarify: by "identify" I meant "to make it easier for a developer/engineer/etc to identify what features the collector corresponds to directly by looking at the NB record"

Reply:
hmm that sounds potentially useful, let me try that